elements/lisk-dpos/src/delegates_list.ts

```diff
@@ -121,26 +171,166 @@ export const getForgerPublicKeysForRound = async (
 	return shuffleDelegateListForRound(round, delegatePublicKeys);
 };
 
+export const isCurrentlyPunished = (
+	height: number,
+	pomHeights: ReadonlyArray<number>,
+): boolean => {
+	if (pomHeights.length === 0) {
+		return false;
+	}
+	const lastPomHeight = Math.max(...pomHeights);
+	if (height - lastPomHeight < PUNISHMENT_PERIOD) {
+		return true;
+	}
+
+	return false;
+};
+
 export class DelegatesList {
 	private readonly rounds: Rounds;
 	private readonly chain: Chain;
 	private readonly activeDelegates: number;
+	private readonly standbyDelegates: number;
+	private readonly voteWeightCapRate: number;
+	private readonly standbyThreshold: bigint;
 	private readonly exceptions: {
 		readonly ignoreDelegateListCacheForRounds?: ReadonlyArray<number>;
 	};
 
 	public constructor({
 		activeDelegates,
+		standbyDelegates,
+		standbyThreshold,
+		voteWeightCapRate,
 		rounds,
 		chain,
 		exceptions,
 	}: DelegatesListConstructor) {
 		this.activeDelegates = activeDelegates;
+		this.standbyDelegates = standbyDelegates;
+		this.standbyThreshold = standbyThreshold;
+		this.voteWeightCapRate = voteWeightCapRate;
 		this.rounds = rounds;
 		this.exceptions = exceptions;
 		this.chain = chain;
 	}
 
+	public async createVoteWeightsSnapshot(
+		height: number,
+		stateStore: StateStore,
+		roundOffset: number = DEFAULT_ROUND_OFFSET,
+	): Promise<void> {
+		const round = this.rounds.calcRound(height) + roundOffset;
+		debug(`Creating vote weight snapshot for round: ${round}`);
+		// This list is before executing the current block in process
+		const originalDelegates = await this.chain.dataAccess.getDelegates();
+
+		// Merge updated delegate accounts
+		const updatedAccounts = stateStore.account.getUpdated();
+		// tslint:disable-next-line readonly-keyword
+		const updatedAccountsMap: { [address: string]: Account } = {};
+		// Convert updated accounts to map for better search
+		for (const account of updatedAccounts) {
+			// Insert only if account is a delegate
+			if (account.username) {
+				updatedAccountsMap[account.address] = account;
+			}
+		}
+		// Inject delegate account if it doesn't exist
+		for (const delegate of originalDelegates) {
+			if (updatedAccountsMap[delegate.address] === undefined) {
+				updatedAccountsMap[delegate.address] = delegate;
+			}
+		}
+		const delegates = [...Object.values(updatedAccountsMap)];
+
+		// Update totalVotesReceived to voteWeight equivalent before sorting
+		for (const account of delegates) {
+			// If the account is being punished, then consider them as vote weight 0
+			if (isCurrentlyPunished(height, account.delegate.pomHeights)) {
+				account.totalVotesReceived = BigInt(0);
+				continue;
+			}
+			const selfVote = account.votes.find(
+				vote => vote.delegateAddress === account.address,
+			);
+			const cappedValue =
+				(selfVote?.amount ?? BigInt(0)) * BigInt(this.voteWeightCapRate);
+			if (account.totalVotesReceived > cappedValue) {
+				account.totalVotesReceived = cappedValue;
+			}
```

Comment:
Why do we update `totalVotesReceived` ? Logicaly, `totalVotesReceived` should not change, but `weight` is re-calculated here.

Reply:
In this selection, account is not set to the state store, so the value itself will not persisted after this function scope.
Therefore, I thought it would be easier to mutate the totalVotesReceived here and sort, then create vote weight entries from there.
We could do it like create the voteWeight list and sort and pick too

Reply:
(not very important though)